File: src/ui/timeline/select.rs
use egui::Pos2;

use crate::{
    project::{Project, timeline::Timeline},
    ui::timeline::{LAYER_HEIGHT, SelectionRect, TimelineWindow},
};

impl TimelineWindow {
    // return true if it was selected
    pub(super) fn handle_select_clip(
        &mut self,
        timeline: &mut Timeline,
        local: Pos2,
        ctrl: bool,
    ) -> bool {
        let Some((_layer_idx, _clip_idxx, clip)) = self.clip_at(&timeline, &local) else {
            self.selected_clips.clear();
            return false;
        };

        if !ctrl {
            self.selected_clips.clear();
        }
        self.selected_clips.insert(clip.id);
        return true;
    }

    pub(super) fn handle_area_sel_start(&mut self, local: Pos2, ctrl: bool) {
        if !ctrl {
            self.selected_clips.clear();
        }
        self.selection_rect = Some(SelectionRect {
            start: local,
            current: local,
        });
    }

    pub(super) fn handle_area_sel_continue(&mut self, local: Pos2) {
        if let Some(sel) = &mut self.selection_rect {
            sel.current = local;
        }
    }

    pub(super) fn handle_area_sel_drop(&mut self, timeline: &mut Timeline) {
        let Some(sel) = &self.selection_rect else {
            return;
        };

        let sel_rect = egui::Rect::from_two_pos(sel.start, sel.current);

        for (layer_idx, layer) in timeline.layers.iter().enumerate() {
            for (_clip_idx, clip) in layer.clips.iter().enumerate() {
                let clip_x_start = self.frame_to_x(clip.position);
                let clip_x_end = self.frame_to_x(clip.position + clip.duration);
                let clip_y_start = self.layer_to_y(layer_idx);
                let clip_y_end = clip_y_start + LAYER_HEIGHT;

                let clip_rect = egui::Rect::from_min_max(
                    egui::pos2(clip_x_start, clip_y_start),
                    egui::pos2(clip_x_end, clip_y_end),
                );

                if sel_rect.intersects(clip_rect) {
                    self.selected_clips.insert(clip.id);
                }
            }
        }

        self.selection_rect = None;
    }
}

File: gui/src/window/widget/timeline/select.cpp
#include "TimelineWidget.h"
#include "wrapper/project/Timeline.h"

namespace esotereel::window {
// return true if selected
bool TimelineWidget::handleSelectClip(const Project &project, const QPoint &mousePos, bool ctrl) {
    auto [clip, layerId] = this->findClipAt(project, mousePos);
    if (!clip.isValid()) {
        if (!ctrl) {
            this->selectedClipIds.clear();
            update();
            return false;
        }
        return false;
    }

    uint64_t id = clip.id();
    if (ctrl) {
        if (this->selectedClipIds.count(id)) {
            this->selectedClipIds.erase(id);
        } else {
            this->selectedClipIds.insert(id);
        }
    } else {
        this->selectedClipIds.clear();
        this->selectedClipIds.insert(id);
    }
    update();
    return true;
}

std::optional<DragAreaSel> TimelineWidget::handleAreaSelStart(const QPoint &mousePos, bool ctrl) {
    if (mousePos.x() <= LABEL_WIDTH || mousePos.y() <= RULER_HEIGHT) {
        return std::nullopt;
    }

    if (!ctrl) {
        this->selectedClipIds.clear();
    }

    update();
    return DragAreaSel{
        mousePos,
        mousePos,
    };
}

void TimelineWidget::handleAreaSelContinue(const QPoint &mousePos) {
    if (auto *sel = std::get_if<DragAreaSel>(&this->dragState)) {
        sel->current = mousePos;
    }
    update();
}

void TimelineWidget::handleAreaSelEnd(const Project &project) {
    auto *sel = std::get_if<DragAreaSel>(&this->dragState);
    if (!sel) {
        return;
    }

    QRect selRect = QRect(sel->start.toPoint(), sel->current.toPoint()).normalized();

    std::unique_ptr<RenderRows> &rr = this->cachedRows;
    if (!rr) {
        return;
    }

    // TODO: optimize
    size_t layerIdx = 0;
    for (auto const &row : rr->rows()) {
        for (auto const &info : rr->clipsFor(row)) {
            double_t clipXStart = this->frameToX(info.abs_frame);
            double_t clipXEnd = this->frameToX(info.abs_frame + info.duration);
            double_t clipYStart = this->rowToY(layerIdx);
            double_t clipYEnd = clipYStart + LAYER_HEIGHT;

            QRect clip_rect(clipXStart, clipYStart, clipXEnd, clipYEnd);

            if (selRect.intersects(clip_rect)) {
                this->selectedClipIds.insert(info.clip_id);
            }
        }
        layerIdx++;
    }

    update();
}

void TimelineWidget::drawSelectionRect(QPainter &p, const QRect &r) const {
    auto *sel = std::get_if<DragAreaSel>(&this->dragState);
    if (!sel) {
        return;
    }

    QRectF selRect(r.topLeft() + sel->start, r.topLeft() + sel->current);

    p.setBrush(QColor(100, 150, 255, 64));
    p.setPen(QPen(QColor(100, 150, 255), 1));
    p.drawRect(selRect);
}
} // namespace esotereel::window